### PR TITLE
SLING-7793 ACLs pipes

### PR DESCRIPTION
**Pull request by dichaudhary**
ACLs pipe to set or retrieve acls on a resource , few example are given below.
**.allow(username/principalName/authorizableId)** (default allow jcr:all on currentResource to userName)
**.allow(username/principalName/authorizableId).with('jcr:privileges',[jcr:read,jcr:write])** sets given allow privileges on currentResource to userName).
**.deny(username/principalName/authorizableId)** (default deny jcr:read on currentResource to userName)
**.deny(username/principalName/authorizableId).with('jcr:privileges',[jcr:read,jcr:write])** sets given deny privileges on currentResource to userName).
**.acls()** will return current resource, with in output bindings array of ACLs mentioning principal, privilege of that resource. 
**.acls()** If current resource is an authorizable, then returns ACLs attached to that user in the repository in bindings.

test case for current resource as an authorizable is marked as ignored since mocking for authorizable is getting failed.

@npeltier

### CODE REVIEW

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 1]
```diff
@@ -177,6 +177,11 @@ protected void bindACLs(Resource resource) {
                 }
                 principal_Privileges_Mappings.add(PN_AUTHORIZABLE, entry.getPrincipal().getName());
                 principal_Privileges_Mappings.add(PRIVILEGES_JSON_KEY, privileges);
```
Comment (dichaudh) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: adding key for whether entry is allow or deny.

**Comment by npeltier** [comment]
@dichaudh haven't looked at it yet, but API seems strange to me : 
- .acls('allow','true') / .acls('deny','true') could be  .acls('allow') / .acls('deny') or (preferred) .allow(expr) .deny(expr),
- ```.acls('authorizable','true')```, i guess .acls() with current resource being an authorizable is enough

**Comment by dichaudh** [comment]
@npeltier 
I already tried with  .acls('allow')  or .acls('deny')  but api expect conf params to be even in number  so went with similar implementation as done in Authorizable pipe -https://sling.apache.org/documentation/bundles/sling-pipes/writers.html

.acls('authorizable','true') can be changed to simply .acls() but I thought of differentiating between the use case when only the acls under authorizable node has to be put or the acls of authorizable on the repository has to be put in bindings based on authorizable is true or not
let me know your thought on it ,
Thanks,
Dinesh

File: pom.xml [hunk 2]
```diff
@@ -248,6 +254,17 @@
       <version>1.3.2</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.sling</groupId>
+      <artifactId>org.apache.sling.testing.sling-mock-oak</artifactId>
```
Comment (npeltier) [inline pom.xml]: pls don't add this if not needed in the end

File: pom.xml [hunk 3]
```diff
@@ -248,6 +254,17 @@
...
+      <version>2.1.0</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-simple</artifactId>
```
Comment (npeltier) [inline pom.xml]: same here

File: src/main/java/org/apache/sling/pipes/PipeBuilder.java [hunk 4]
```diff
@@ -252,6 +253,16 @@
...
 
+    /**
+     * attach an ACL pipe to the current context
+     * @param conf configuration parameters
+     * @return updated instance of PipeBuilder
+     * @throws IllegalAccessException in case it's called with bad configuration
+     */
+    @PipeExecutor(command = "acls", resourceType = ACLPipe.RESOURCE_TYPE, pipeClass = ACLPipe.class,
+            description = "sets acls on the resource or output each acls on the resource or  acls for authorizable in repository in bindings")
+    PipeBuilder acls(Object... conf) throws IllegalAccessException;
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/PipeBuilder.java]: i think that the following is more appropriate as discussed:

- allow(String expr),
- deny(String expr),
- acls()

allow or deny will internally set your allow & deny flag

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+            authorizable = properties.get(PN_AUTHORIZABLE, false);
+        }
+    }
+
+    @Override
+    public Iterator<Resource> computeOutput() throws Exception {
+        Resource resource = getInput();
+        if (resource != null) {
+            if (allow || deny) {
+                logger.info("Going to changing ACL for the resource at path {}", resource.getPath());
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: debug would be better here i think

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                if (StringUtils.isEmpty(getExpr())) {
+                    throw new Exception("expression for the principal or authorizable Id should be provided or provided correctly for privileges to be set");
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: may be IllegalArgumentException ?

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                }
+                Principal principal = getPrincipalFor(getExpr());
+                if ( null != privilegesInput && privilegesInput.length == 0 ){
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: pls use `StringUtils.isBlank`

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                    // create a privilege set with jcr:all
+                    privileges = allow ? AccessControlUtils.privilegesFromNames(session, Privilege.JCR_ALL) : AccessControlUtils.privilegesFromNames(session, Privilege.JCR_READ);
+                }
+                else {
+                    try {
+                        privilegesInput = processPrivilegesInput(privilegesInput);
+                        privileges = AccessControlUtils.privilegesFromNames(session, privilegesInput);
+                    }
+                    catch (Exception e) {
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: i guess you don't want things to be caught, here right? if an exception happens it's better to get out of here. please either remove the catch, either re-throw

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+                        logger.error("unable to get privileges , either privileges input is wrong either error evaluting bindings for privileges", e);
+                    }
+                }
+                if (!isDryRun()) {
+                    if (allow) {
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: i would not do a test just for the logging, you can just output privileges and some flag for the allow/deny part. Do log work with more than 2 {}?

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+                        logger.info("adding privileges {} for principal {} allow {}",ArrayUtils.toString(privileges), principal.getName(), allow);
+                        AccessControlUtils.addAccessControlEntry(session, resource.getPath(), principal, privileges, true);
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: pls move this in a single method, and just englob that line in the dryRun test: it's important we have the logging of what it's supposed to do at that point

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+                    } else if (deny) {
+                        logger.info("adding privileges {} for principal {} deny {}",ArrayUtils.toString(privileges), principal.getName(), deny);
+                        AccessControlUtils.addAccessControlEntry(session, resource.getPath(), principal, privileges, false);
+                    }
+                    if (session.hasPendingChanges()) {
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: if you can, please not save here, plumber will do that for you and saving less will be more performant

File: src/main/java/org/apache/sling/pipes/internal/ACLPipe.java [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+                if (entry.isAllow()) {
+                    principal_Privileges_Mappings.add(PN_ALLOW, true);
+                } else {
+                    principal_Privileges_Mappings.add(PN_DENY, true);
+                }
+                principal_Privileges_Array.add(principal_Privileges_Mappings);
+            }
+            outputBinding = JsonUtil.toString(principal_Privileges_Array);
+        } catch ( Exception e ) {
+            logger.error("unable to bind acls", e);
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/internal/ACLPipe.java]: may be in that case, we need empty output bindings

File: src/test/java/org/apache/sling/pipes/internal/ACLPipeTest.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonStructure;
+import java.security.Principal;
+import java.util.Iterator;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class ACLPipeTest extends AbstractPipeTest {
```
Comment (npeltier) [inline src/test/java/org/apache/sling/pipes/internal/ACLPipeTest.java]: pls add @Ignore to tests that break the build

**Comment by dichaudh** [comment]
@npeltier  please find my changes done with latest commit .
Thanks
Dinesh

File: src/main/java/org/apache/sling/pipes/PipeBuilder.java [hunk 14]
```diff
@@ -252,6 +253,36 @@
...
 
+    /**
+     * attach an ACL pipe to the current context
+     * @param conf configuration parameters
+     * @return updated instance of PipeBuilder
+     * @throws IllegalAccessException in case it's called with bad configuration
+     */
+    @PipeExecutor(command = "acls", resourceType = ACLPipe.RESOURCE_TYPE, pipeClass = ACLPipe.class,
+            description = "output each acls on the resource or  acls for authorizable in repository in bindings")
+    PipeBuilder acls(Object... conf) throws IllegalAccessException;
```
[the remark this answers is not in the document]
Comment (dichaudh) [reply src/main/java/org/apache/sling/pipes/PipeBuilder.java]: correct!

File: src/main/java/org/apache/sling/pipes/PipeBuilder.java [hunk 15]
```diff
@@ -252,6 +253,36 @@
...
+     * @return updated instance of PipeBuilder
+     * @throws IllegalAccessException in case it's called with bad configuration
+     */
+    @PipeExecutor(command = "allow", resourceType = ACLPipe.RESOURCE_TYPE, pipeClass = ACLPipe.class,
+            description = "sets allow acls on the resource")
+    PipeBuilder allow(String expr) throws IllegalAccessException;
+
+    /**
+     * attach an ACL pipe to the current context and sets deny acls on the resource
+     * @param conf configuration parameters
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/PipeBuilder.java]: please remove absent parameter from javadoc here

File: src/main/java/org/apache/sling/pipes/PipeBuilder.java [hunk 16]
```diff
@@ -252,6 +253,36 @@
...
+     * @return updated instance of PipeBuilder
+     * @throws IllegalAccessException in case it's called with bad configuration
+     */
+    @PipeExecutor(command = "acls", resourceType = ACLPipe.RESOURCE_TYPE, pipeClass = ACLPipe.class,
+            description = "output each acls on the resource or  acls for authorizable in repository in bindings")
+    PipeBuilder acls(Object... conf) throws IllegalAccessException;
+
+    /**
+     * attach an ACL pipe to the current context and sets allow acls on the resource
+     * @param conf configuration parameters
```
Comment (npeltier) [inline src/main/java/org/apache/sling/pipes/PipeBuilder.java]: please remove absent parameter from javadoc here

File: src/test/java/org/apache/sling/pipes/internal/ACLPipeTest.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
+package org.apache.sling.pipes.internal;
```
Comment (npeltier) [inline src/test/java/org/apache/sling/pipes/internal/ACLPipeTest.java]: also @dichaudh you miss a license here, which make rat plugin fails (try `mvn clean install`)

Comment (dichaudh) [reply src/test/java/org/apache/sling/pipes/internal/ACLPipeTest.java]: Thanks @npeltier  for the feedback , raised a new commit for the changes

File: pom.xml [hunk 18]
```diff
@@ -114,7 +114,7 @@
             <groupId>*</groupId>
             <artifactId>*</artifactId>
         </exclusion>
-    </exclusions>
+      </exclusions>
```
Comment (npeltier) [inline pom.xml]: pls remove that change

Comment (dichaudh) [reply pom.xml]: @npeltier  unable to see the ident or trailing space problem on my local

Comment (npeltier) [reply pom.xml]: right, me neither, let's move on

**Comment by dichaudh** [comment]
Raising a new one  as there are too many commits and squashing has failed 
refer : https://github.com/apache/sling-org-apache-sling-pipes/pull/7

**Pull request closed without merging**